### PR TITLE
feat(masked errors): rebuild temporary repair as share-able permanent gateway errorhandler

### PR DESCRIPTION
**Pull request by cmharlow**
## Goal

**Changed scope: this now addresses a new work requirement surfaced in slack discussions - the ability for multiple gateways to use the same bug repair. This PR is updated to in theory support that new request. The original bug tickets' scope has been completed already.**

Repair gateway-specific masking errors and add some "integration" tests that reflect the HTTP-based, Javascript class-stripping issues of passing our custom Errors from the subgraph to the gateway.

The repair (hopefully temporary) is to simply have an additional if/else path that acts on GraphQL type & the extensions.code value being something we don't want masked.

Note: making PR and pushing tests alone first to confirm the capture the bad behavior of the gateway, then pushing a possible repair. You can see the 'internal error' masking occurring in this test run: https://github.com/Pocket/apollo-utils/actions/runs/4365108384/jobs/7633288751 (had to alter some of the jest config to forceexit since the http servers aren't closing cleanly - entirely open to ideas on how to better repair that).

Issues:
* https://getpocket.atlassian.net/browse/INFRA-1041

## Notes:

First, the heart of the issue is that Gateway is making HTTP requests to Subgraphs & getting JSON response back, which loses javascript context on errors.

This makes the request: https://github.com/apollographql/federation/blob/4f6d7bde74c537ea85d0a5435986bedf29a21e7f/gateway-js/src/executeQueryPlan.ts#L481
This checks for errors in the request JSON: https://github.com/apollographql/federation/blob/4f6d7bde74c537ea85d0a5435986bedf29a21e7f/gateway-js/src/executeQueryPlan.ts#L494
This does some further SubGraph Error handling: https://github.com/apollographql/federation/blob/4f6d7bde74c537ea85d0a5435986bedf29a21e7f/gateway-js/src/executeQueryPlan.ts#L882
In particular, see this relevant to us note about error types: https://github.com/apollographql/federation/blob/4f6d7bde74c537ea85d0a5435986bedf29a21e7f/gateway-js/src/executeQueryPlan.ts#L905
where our JSON response gets cast as a generic GraphQLError: https://github.com/apollographql/federation/blob/4f6d7bde74c537ea85d0a5435986bedf29a21e7f/gateway-js/src/executeQueryPlan.ts#L909

In order to fully test this issue, I ran gateway and a few services (including shareable lists) locally. Here are some findings of the request & error responses being passed between the two:

#### Shareable Lists API

```
RECV <<< POST / localhost {
  query: 'mutation Mutation__shareable_lists_api__0($data:CreateShareableListInput!){createShareableList(data:$data){title description status moderationStatus}}',
  variables: {
    data: {
      title: 'charlow-test5',
      description: 'making test public collections'
    }
  },
  operationName: 'Mutation__shareable_lists_api__0'
}
SEND >>> 200 {"errors":[{"message":"A list with the title \"charlow-test5\" already exists","locations":[{"line":1,"column":76}],"path":["createShareableList"],"extensions":{"code":"BAD_USER_INPUT","stacktrace":["UserInputError: A list with the title \"charlow-test5\" already exists","    at new CustomGraphQLError (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/node_modules/@pocket-tools/apollo-utils/dist/errorHandler/errorHandler.js:30:9)","    at new UserInputError (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/node_modules/@pocket-tools/apollo-utils/dist/errorHandler/errorHandler.js:71:9)","    at createShareableList (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/dist/database/mutations/ShareableList.js:19:15)","    at async executeMutation (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/dist/public/resolvers/mutations/utils.js:10:12)","    at async createShareableList (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/dist/public/resolvers/mutations/ShareableList.js:7:12)"]}}],"data":{"createShareableList":null}}
```

#### Gateway / Client API

```
RECV <<< POST / localhost {
  operationName: 'Mutation',
  variables: {
    data: {
      title: 'charlow-test5',
      description: 'making test public collections'
    }
  },
  query: 'mutation Mutation($data: CreateShareableListInput!) {\n' +
    '  createShareableList(data: $data) {\n' +
    '    title\n' +
    '    description\n' +
    '    status\n' +
    '    moderationStatus\n' +
    '  }\n' +
    '}\n'
}
SEND >>> 200 {"errors":[{"message":"Internal server error","path":["createShareableList"],"extensions":{"code":"INTERNAL_SERVER_ERROR"}}],"data":{"createShareableList":null}}
```

#### Shareable Lists API

```
info: { message:{"id":"a79af572-b89a-43c8-acef-139c6a60dc59","event":"request","operationName":"Mutation__shareable_lists_api__0","query":"mutation Mutation__shareable_lists_api__0($data:CreateShareableListInput!){createShareableList(data:$data){title description status moderationStatus}}","variables":{"data":{"title":"charlow-test5","description":"making test public collections"}}}, timestamp: 2023-03-06 20:40:23:4023 }
debug: { message:{"id":"a79af572-b89a-43c8-acef-139c6a60dc59","event":"parsingDidStart"}, timestamp: 2023-03-06 20:40:23:4023 }
debug: { message:{"id":"a79af572-b89a-43c8-acef-139c6a60dc59","event":"validationDidStart"}, timestamp: 2023-03-06 20:40:23:4023 }
debug: { message:{"id":"a79af572-b89a-43c8-acef-139c6a60dc59","event":"didResolveOperation","operationName":"Mutation__shareable_lists_api__0"}, timestamp: 2023-03-06 20:40:23:4023 }
debug: { message:{"id":"a79af572-b89a-43c8-acef-139c6a60dc59","event":"responseForOperation","response":{"http":{"headers":{}}}}, timestamp: 2023-03-06 20:40:23:4023 }
debug: { message:{"id":"a79af572-b89a-43c8-acef-139c6a60dc59","event":"executionDidStart"}, timestamp: 2023-03-06 20:40:23:4023 }
error: { message:{"id":"a79af572-b89a-43c8-acef-139c6a60dc59","event":"errors","errors":[{"message":"A list with the title \"charlow-test5\" already exists","locations":[{"line":1,"column":76}],"path":["createShareableList"],"extensions":{"code":"BAD_USER_INPUT"}}]}, timestamp: 2023-03-06 20:40:23:4023 }
debug: { message:{"id":"a79af572-b89a-43c8-acef-139c6a60dc59","event":"response"}, timestamp: 2023-03-06 20:40:23:4023 }
```

#### Client API

```
info: { message:{"id":"8dffe46b-4201-4074-983a-42b7c6de9243","event":"request","operationName":"Mutation","query":"mutation Mutation($data: CreateShareableListInput!) { createShareableList(data: $data) { title description status moderationStatus } } ","variables":{"data":{"title":"charlow-test5","description":"making test public collections"}}}, timestamp: 2023-03-06 20:40:23:4023 }
debug: { message:{"id":"8dffe46b-4201-4074-983a-42b7c6de9243","event":"parsingDidStart"}, timestamp: 2023-03-06 20:40:23:4023 }
debug: { message:{"id":"8dffe46b-4201-4074-983a-42b7c6de9243","event":"validationDidStart"}, timestamp: 2023-03-06 20:40:23:4023 }
debug: { message:{"id":"8dffe46b-4201-4074-983a-42b7c6de9243","event":"didResolveOperation","operationName":"Mutation"}, timestamp: 2023-03-06 20:40:23:4023 }
debug: { message:{"id":"8dffe46b-4201-4074-983a-42b7c6de9243","event":"responseForOperation","response":{"http":{"headers":{}}}}, timestamp: 2023-03-06 20:40:23:4023 }
debug: { message:{"id":"8dffe46b-4201-4074-983a-42b7c6de9243","event":"executionDidStart"}, timestamp: 2023-03-06 20:40:23:4023 }
error: { message:{"id":"8dffe46b-4201-4074-983a-42b7c6de9243","event":"errors","errors":[{"message":"A list with the title \"charlow-test5\" already exists","path":["createShareableList"],"extensions":{"code":"BAD_USER_INPUT","stacktrace":["UserInputError: A list with the title \"charlow-test5\" already exists","    at new CustomGraphQLError (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/node_modules/@pocket-tools/apollo-utils/dist/errorHandler/errorHandler.js:30:9)","    at new UserInputError (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/node_modules/@pocket-tools/apollo-utils/dist/errorHandler/errorHandler.js:71:9)","    at createShareableList (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/dist/database/mutations/ShareableList.js:19:15)","    at async executeMutation (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/dist/public/resolvers/mutations/utils.js:10:12)","    at async createShareableList (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/dist/public/resolvers/mutations/ShareableList.js:7:12)"],"serviceName":"shareable-lists-api"}}]}, timestamp: 2023-03-06 20:40:23:4023 }
debug: { message:{"id":"8dffe46b-4201-4074-983a-42b7c6de9243","event":"response"}, timestamp: 2023-03-06 20:40:23:4023 }
info: { message:{"id":"da508cdb-2183-489f-ba9e-09ffa828a2a0","event":"request","operationName":"IntrospectionQuery","query":"query IntrospectionQuery { __schema { queryType { name } mutationType { name } subscriptionType { name } types { ...FullType } directives { name description locations args { ...InputValue } } } } fragment FullType on __Type { kind name description fields(includeDeprecated: true) { name description args { ...InputValue } type { ...TypeRef } isDeprecated deprecationReason } inputFields { ...InputValue } interfaces { ...TypeRef } enumValues(includeDeprecated: true) { name description isDeprecated deprecationReason } possibleTypes { ...TypeRef } } fragment InputValue on __InputValue { name description type { ...TypeRef } defaultValue } fragment TypeRef on __Type { kind name ofType { kind name ofType { kind name ofType { kind name ofType { kind name ofType { kind name ofType { kind name ofType { kind name } } } } } } } } ","variables":{}}, timestamp: 2023-03-06 20:40:23:4023 }
```

#### Shareable Lists API

formattedError value, type, & class instance: 
```
{
  message: 'A list with the title "charlow-test5" already exists',
  locations: [ { line: 1, column: 76 } ],
  path: [ 'createShareableList' ],
  extensions: {
    code: 'BAD_USER_INPUT',
    stacktrace: [
      'UserInputError: A list with the title "charlow-test5" already exists',
      '    at new CustomGraphQLError (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/node_modules/@pocket-tools/apollo-utils/dist/errorHandler/errorHandler.js:30:9)',
      '    at new UserInputError (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/node_modules/@pocket-tools/apollo-utils/dist/errorHandler/errorHandler.js:71:9)',
      '    at createShareableList (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/dist/database/mutations/ShareableList.js:19:15)',
      '    at async executeMutation (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/dist/public/resolvers/mutations/utils.js:10:12)',
      '    at async createShareableList (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/dist/public/resolvers/mutations/ShareableList.js:7:12)'
    ]
  }
}
object
Object
```

Error value, type, & class instance: 
```
UserInputError: A list with the title "charlow-test5" already exists
    at new CustomGraphQLError (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/node_modules/@pocket-tools/apollo-utils/dist/errorHandler/errorHandler.js:30:9)
    at new UserInputError (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/node_modules/@pocket-tools/apollo-utils/dist/errorHandler/errorHandler.js:71:9)
    at createShareableList (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/dist/database/mutations/ShareableList.js:19:15)
    at async executeMutation (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/dist/public/resolvers/mutations/utils.js:10:12)
    at async createShareableList (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/dist/public/resolvers/mutations/ShareableList.js:7:12) {
  path: [ 'createShareableList' ],
  locations: [ { line: 1, column: 76 } ],
  extensions: { code: 'BAD_USER_INPUT' }
}
object
GraphQLError
```

#### Client API / Gateway

formattedError value, type, & class instance: 
```
{
  message: 'A list with the title "charlow-test5" already exists',
  path: [ 'createShareableList' ],
  extensions: {
    code: 'BAD_USER_INPUT',
    stacktrace: [
      'GraphQLError: A list with the title "charlow-test5" already exists',
      '    at downstreamServiceError (/Users/charlow/Work/src/github.com/pocket/client-api/node_modules/@apollo/gateway/src/executeQueryPlan.ts:909:12)',
      '    at /Users/charlow/Work/src/github.com/pocket/client-api/node_modules/@apollo/gateway/src/executeQueryPlan.ts:498:9',
      '    at Array.map (<anonymous>)',
      '    at sendOperation (/Users/charlow/Work/src/github.com/pocket/client-api/node_modules/@apollo/gateway/src/executeQueryPlan.ts:497:38)',
      '    at processTicksAndRejections (node:internal/process/task_queues:96:5)',
      '    at async /Users/charlow/Work/src/github.com/pocket/client-api/node_modules/@apollo/gateway/src/executeQueryPlan.ts:378:41',
      '    at async executeNode (/Users/charlow/Work/src/github.com/pocket/client-api/node_modules/@apollo/gateway/src/executeQueryPlan.ts:318:9)',
      '    at async /Users/charlow/Work/src/github.com/pocket/client-api/node_modules/@apollo/gateway/src/executeQueryPlan.ts:133:27',
      '    at async /Users/charlow/Work/src/github.com/pocket/client-api/node_modules/@apollo/gateway/src/index.ts:820:28',
      '    at async execute (/Users/charlow/Work/src/github.com/pocket/client-api/node_modules/@apollo/server/src/requestPipeline.ts:520:22)'
    ],
    serviceName: 'shareable-lists-api'
  }
}
object
Object
```

Error value, type, & class instance: 
```
GraphQLError: A list with the title "charlow-test5" already exists
    at downstreamServiceError (/Users/charlow/Work/src/github.com/pocket/client-api/node_modules/@apollo/gateway/src/executeQueryPlan.ts:909:12)
    at /Users/charlow/Work/src/github.com/pocket/client-api/node_modules/@apollo/gateway/src/executeQueryPlan.ts:498:9
    at Array.map (<anonymous>)
    at sendOperation (/Users/charlow/Work/src/github.com/pocket/client-api/node_modules/@apollo/gateway/src/executeQueryPlan.ts:497:38)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async /Users/charlow/Work/src/github.com/pocket/client-api/node_modules/@apollo/gateway/src/executeQueryPlan.ts:378:41
    at async executeNode (/Users/charlow/Work/src/github.com/pocket/client-api/node_modules/@apollo/gateway/src/executeQueryPlan.ts:318:9)
    at async /Users/charlow/Work/src/github.com/pocket/client-api/node_modules/@apollo/gateway/src/executeQueryPlan.ts:133:27
    at async /Users/charlow/Work/src/github.com/pocket/client-api/node_modules/@apollo/gateway/src/index.ts:820:28
    at async execute (/Users/charlow/Work/src/github.com/pocket/client-api/node_modules/@apollo/server/src/requestPipeline.ts:520:22) {
  path: [ 'createShareableList' ],
  locations: undefined,
  extensions: {
    code: 'BAD_USER_INPUT',
    stacktrace: [
      'UserInputError: A list with the title "charlow-test5" already exists',
      '    at new CustomGraphQLError (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/node_modules/@pocket-tools/apollo-utils/dist/errorHandler/errorHandler.js:30:9)',
      '    at new UserInputError (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/node_modules/@pocket-tools/apollo-utils/dist/errorHandler/errorHandler.js:71:9)',
      '    at createShareableList (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/dist/database/mutations/ShareableList.js:19:15)',
      '    at async executeMutation (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/dist/public/resolvers/mutations/utils.js:10:12)',
      '    at async createShareableList (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/dist/public/resolvers/mutations/ShareableList.js:7:12)'
    ],
    serviceName: 'shareable-lists-api'
  }
}
object
GraphQLError
```

Unwrapped Error value, type, & class instance: 
```
{
  message: 'A list with the title "charlow-test5" already exists',
  locations: [ { line: 1, column: 76 } ],
  path: [ 'createShareableList' ],
  extensions: {
    code: 'BAD_USER_INPUT',
    stacktrace: [
      'UserInputError: A list with the title "charlow-test5" already exists',
      '    at new CustomGraphQLError (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/node_modules/@pocket-tools/apollo-utils/dist/errorHandler/errorHandler.js:30:9)',
      '    at new UserInputError (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/node_modules/@pocket-tools/apollo-utils/dist/errorHandler/errorHandler.js:71:9)',
      '    at createShareableList (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/dist/database/mutations/ShareableList.js:19:15)',
      '    at async executeMutation (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/dist/public/resolvers/mutations/utils.js:10:12)',
      '    at async createShareableList (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/dist/public/resolvers/mutations/ShareableList.js:7:12)'
    ]
  }
}
object
Object
```

Double Unwrapped Error value, type, & class instance: 
```
{
  message: 'A list with the title "charlow-test5" already exists',
  locations: [ { line: 1, column: 76 } ],
  path: [ 'createShareableList' ],
  extensions: {
    code: 'BAD_USER_INPUT',
    stacktrace: [
      'UserInputError: A list with the title "charlow-test5" already exists',
      '    at new CustomGraphQLError (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/node_modules/@pocket-tools/apollo-utils/dist/errorHandler/errorHandler.js:30:9)',
      '    at new UserInputError (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/node_modules/@pocket-tools/apollo-utils/dist/errorHandler/errorHandler.js:71:9)',
      '    at createShareableList (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/dist/database/mutations/ShareableList.js:19:15)',
      '    at async executeMutation (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/dist/public/resolvers/mutations/utils.js:10:12)',
      '    at async createShareableList (/Users/charlow/Work/src/github.com/pocket/shareable-lists-api/dist/public/resolvers/mutations/ShareableList.js:7:12)'
    ]
  }
}
object
Object
```

### CODE REVIEW

**Review by hyperparabolic — CHANGES_REQUESTED** [review]
Great analysis. Super thorough. Test coverage looks great, too.

One tiny implementation nit.

File: src/errorHandler/errorHandler.ts [hunk 1]
```diff
@@ -32,6 +47,11 @@ export function errorHandler(
     // Keep GraphQL errors intact
     // e.g. failed parsing, bad input
     return formattedError;
+  } else if (
+    isGatewayError(error) &&
+    gatewayUnmaskedErrors.indexOf(error.extensions.code as string) > -1
```
[the remark this answers is not in the document]
Comment (cmharlow) [reply src/errorHandler/errorHandler.ts:52]: yeah so trying to implement this reminded why I got to where I got to - I want to pass through a String, and if I just make a Set, it doesn't register as a Set of Strings, but as `const gatewayUnmaskedErrors: Set<InternalErrorCode | ApolloServerErrorCode>` - which blocks trying to find the extension code in it.

File: src/errorHandler/errorHandler.ts [hunk 2]
```diff
@@ -18,6 +21,18 @@ export enum InternalErrorCode {
...
+const isGatewayError = (value: unknown): value is GraphQLError =>
+  !!value &&
+  typeof value == 'object' &&
+  'extensions' in value &&
+  typeof (value as GraphQLError).extensions === 'object' &&
+  'code' in (value as GraphQLError).extensions &&
+  typeof (value as GraphQLError).extensions.code === 'string';
+const pocketErrorCodes = Object.values(InternalErrorCode);
+const apolloErrorCodes = Object.values(ApolloServerErrorCode);
+const gatewayUnmaskedErrors: string[] = [
```
[the remark this answers is not in the document]
Comment (cmharlow) [reply src/errorHandler/errorHandler.ts]: i attempted a version of what i think you proposed here, though its not terribly different from the original in order to make string matching work.

**Review by jpetto — DISMISSED** [review]
this looks good to me!

**Comment by cmharlow** [comment]
closed in favor of https://github.com/Pocket/client-api/pull/348 as the temporary solution at the time.

**Pull request closed without merging**